impexp-core/src/main/java/org/citydb/citygml/exporter/database/content/CityGMLExportManager.java

```diff
@@ -258,6 +260,9 @@ private AbstractGML processObject(AbstractGML object) throws CityGMLExportExcept
 				getExporter(DBLocalAppearance.class).triggerLazyTextureExport(feature);
 		}
 
+		// clear local geometry cache
+		localGeometryCache.clear();
```

Comment:
The method `processObject` is only intended to be invoked for top-level objects. Besides clearing the local geometry cache, also things like executing the export batches or removing filtered LoDs are done here, which are also supposed to be executed only once for every top-level object. 

`processObject` is used by the public methods `exportObject` and `exportAsGlobalFeature`, only the latter of which is part of the ADE API (and thus intended to be used in external code). Yet, we could rename `exportObject` to `exportTopLevelObject` to make its intended use more obvious.

Do you have a test scenario in which `processObject` is used for non-top-level objects?

Reply:
Yes, my update-tool currently uses the method `exportObject` for exporting non-top-level features. I think, renaming the method `exportObject` to `exportTopLevelObject` is not appropriate. We could think about introducing a new method `exportTopLevelObject`, which should first export object and then do additional things (e.g. clearing  local geometry cache) only for top-level feature.

Reply:
Ok, thanks. Please provide a patch file with your suggestions.

Reply:
Please check this patch file [added_exportTopLevelObject_method.zip](https://github.com/3dcitydb/importer-exporter/files/5747756/added_exportTopLevelObject_method.zip)

Reply:
Thanks for the patch file. Hm, the issue with this proposal is that the local geometry cache is never cleared when only `exportObject` is invoked. I am also not sure why `cleanupCityObjects`, `cleanupAppearances`,  `cacheGeometryIds` and export plugins should be skipped?

When looking at your initial comment, what exactly is the problem if the local geometry cache is cleared in `exportObject` when invoking the method for a non-top-level object? You still get back your object.

Reply:
Since the `cleanupCityObjects`, `cleanupAppearances`, `cacheGeometryIds` and export plugins are meant for top-level or global features, They should be skipped in `exportObject` in order to avoid side effects when exporting non-top-level objects. 

You are right, in my proposal, the local geometry cache is never cleared if no top-level or global feature is exported. The local geometry cache should also be cleared in `exportObject`. So, your current implementation is correct and has no problem. 

Another problem raised from this discussion is that we need a clean API allowing for export non-top-level objects without any side effects. I think my patch file is a good starting point. We could tweak it in a separate PR.

Reply:
I agree, @yaozhihang. The `CityGMLExportManager` class is not designed as a general-purpose API for exporting any type of city object in any context. It is designed as helper class specifically for the CityGML export operation of the Importer/Exporter and, thus, might not be appropriate in other use cases or tools. For example, the public API for ADE extensions is `CityGMLExportHelper` (not `CityGMLExportManager`) and intentionally lacks the `exportObject` method.

I also think that we need a more general API which would allow implementers to easily read city objects from the database for further processing. We should carefully define this API and check whether it can be built based on the `CityGMLExportManager` class or whether we should put it into its own implementation.

I would also prefer to decouple this discussion from this PR.

Reply:
Agree.